src/db.rs

```diff
@@ -146,25 +146,65 @@ pub fn connection_url(url: &str) -> String {
 
 pub trait RequestTransaction {
     /// Obtain a read/write database connection from the primary pool
-    fn db_conn(&self) -> Result<DieselPooledConn<'_>, PoolError>;
+    fn db_write(&self) -> Result<DieselPooledConn<'_>, PoolError>;
 
     /// Obtain a readonly database connection from the replica pool
     ///
     /// If there is no replica pool, the primary pool is used instead.
-    fn db_read_only(&self) -> Result<DieselPooledConn<'_>, PoolError>;
+    fn db_read(&self) -> Result<DieselPooledConn<'_>, PoolError>;
+
+    /// Obtain a readonly database connection from the primary pool
+    ///
+    /// If the primary pool is unavailable, the replica pool is used instead.
+    fn db_read_prefer_primary(&self) -> Result<DieselPooledConn<'_>, PoolError>;
 }
 
 impl<T: RequestExt + ?Sized> RequestTransaction for T {
-    fn db_conn(&self) -> Result<DieselPooledConn<'_>, PoolError> {
+    fn db_write(&self) -> Result<DieselPooledConn<'_>, PoolError> {
         self.app().primary_database.get()
     }
 
-    fn db_read_only(&self) -> Result<DieselPooledConn<'_>, PoolError> {
-        match &self.app().read_only_replica_database {
-            Some(pool) => pool.get(),
+    fn db_read(&self) -> Result<DieselPooledConn<'_>, PoolError> {
+        let read_only_pool = self.app().read_only_replica_database.as_ref();
+        match read_only_pool.map(|pool| pool.get()) {
+            // Replica is available
+            Some(Ok(connection)) => Ok(connection),
+
+            // Replica failed, but primary might be available
+            Some(Err(error)) => {
```

Comment:
I'm tempted to say we should failover only when `PoolError::UnhealthyPool` is returned (the amount of connections in the database is 0) instead of any error, otherwise we risk losing visibility into "smaller" bursts of errors.